exporter/googlecloudpubsubexporter/exporter.go

```diff
@@ -16,54 +16,127 @@ package googlecloudpubsubexporter
 
 import (
 	"context"
+	"fmt"
 
+	pubsub "cloud.google.com/go/pubsub/apiv1"
+	"github.com/google/uuid"
 	"go.opentelemetry.io/collector/component"
-	"go.opentelemetry.io/collector/consumer"
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.uber.org/zap"
+	"google.golang.org/api/option"
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+	"google.golang.org/grpc"
 )
 
 const name = "googlecloudpubsub"
 
 type pubsubExporter struct {
 	instanceName string
 	logger       *zap.Logger
-
-	topicName string
-
-	//
-	userAgent string
-	ceSource  string
-	config    *Config
-	//
+	topicName    string
+	client       *pubsub.PublisherClient
+	cancel       context.CancelFunc
+	userAgent    string
+	ceSource     string
+	config       *Config
 }
 
 func (*pubsubExporter) Name() string {
 	return name
 }
 
+type Encoding int
+
+const (
+	OtlpProtoTrace  = iota
+	OtlpProtoMetric = iota
+	OtlpProtoLog    = iota
+)
+
 func (ex *pubsubExporter) start(ctx context.Context, _ component.Host) error {
+	ctx, ex.cancel = context.WithCancel(ctx)
+
+	if ex.client == nil {
+		copts := ex.generateClientOptions()
+		client, err := pubsub.NewPublisherClient(ctx, copts...)
+		if err != nil {
+			return fmt.Errorf("failed creating the gRPC client to Pubsub: %w", err)
+		}
+
+		ex.client = client
+	}
 	return nil
 }
 
 func (ex *pubsubExporter) shutdown(context.Context) error {
+	if ex.client != nil {
+		ex.client.Close()
+		ex.client = nil
+	}
 	return nil
 }
 
-func (ex *pubsubExporter) Capabilities() consumer.Capabilities {
-	return consumer.Capabilities{
-		MutatesData: false,
+func (ex *pubsubExporter) generateClientOptions() []option.ClientOption {
+	var copts []option.ClientOption
+	if ex.userAgent != "" {
+		copts = append(copts, option.WithUserAgent(ex.userAgent))
 	}
+	if ex.config.Endpoint != "" {
+		if ex.config.Insecure {
+			var dialOpts []grpc.DialOption
+			if ex.userAgent != "" {
+				dialOpts = append(dialOpts, grpc.WithUserAgent(ex.userAgent))
+			}
+			conn, _ := grpc.Dial(ex.config.Endpoint, append(dialOpts, grpc.WithInsecure())...)
+			copts = append(copts, option.WithGRPCConn(conn))
+		} else {
+			copts = append(copts, option.WithEndpoint(ex.config.Endpoint))
+		}
+	}
+	return copts
+}
+
+func (ex *pubsubExporter) publishMessage(ctx context.Context, encoding Encoding, data []byte) error {
+	id, _ := uuid.NewRandom()
+	attributes := map[string]string{
+		"ce-specversion": "1.0",
+		"ce-id":          id.String(),
+		"ce-source":      ex.ceSource,
+	}
+	switch encoding {
+	case OtlpProtoTrace:
+		attributes["ce-type"] = "org.opentelemetry.otlp.traces.v1"
+		attributes["ce-datacontenttype"] = "application/x-protobuf"
+	case OtlpProtoMetric:
+		attributes["ce-type"] = "org.opentelemetry.otlp.metrics.v1"
+		attributes["ce-datacontenttype"] = "application/x-protobuf"
+	case OtlpProtoLog:
+		attributes["ce-type"] = "org.opentelemetry.otlp.logs.v1"
+		attributes["ce-datacontenttype"] = "application/x-protobuf"
+	}
```

Comment:
Have you considered compressing the data with gzip here? We have a fairly similar implementation at Shopify. Gzipping is a fairly trivial addition, although I don't know whether or how that's supported with CloudEvents. Our implementation uses common HTTP headers as attributes:
```
			"Content-Encoding": "gzip",
			"Content-Type":     "application/protobuf",
			"Format":           "otlp",
```

Reply:
I did, looking at the volume we're pushing it could start being a cost-saving in the long run. But I would **do this as an iteration**, first the simple groundwork. I also would like to add the new Mashallers, again as an iteration. The above implementation already has proven stable for months.

For the CloudEvent spec, this is tricky. I'll actually reach out to how this could fit into the spec. (Why CloudEvents? It a good and simple spec, laying out simple rules. In theory, if you would route this over a CloudEvent system it would work out of the box).

Reply:
> The above implementation already has proven stable for months.

Fair. By the same token, we've been running with compression in production for well over a year. Ideally, we'd like to switch to your exporter and remove some more custom code from our collector, but this is likely a blocker for us. No pressure, though 😄 - we'll wait.

Reply:
I've re-read the CloudEvent spec (more specifically the Pubsub version of it), and it says: if "content-type" is present, it MUST be used as the "datacontenttype", so I've changed it to "content-type": "application/protobuf" without the x-. So now it's aligned with your implementation. I did use the spec's, all lower-case though. https://github.com/google/knative-gcp/blob/main/docs/spec/pubsub-protocol-binding.md#311-content-type